### PR TITLE
ES6 it

### PR DESCRIPTION
**Pull request by rogierslag**
@inventid/tech This might have gone slightly out of hand. Sorry for that!

Anyhow: I really wanted to ES6 this since its the only location in stack where we were using Javascript like it 1995 and that really bothered me. Apart from that this allows us to do later refactorings better. And I renamed some files etc

While I was add it, I improved Docker, ditched Vagrant, removed JSHint, finished eslint and updated all dependencies

Ready for review!

Todo:

- [x] Drop the `dist/` folder from the repo
- [x] Rewrite all to ES6 syntax
- [x] Test Docker container
- [x] Ditch the vagrant stuff
- [x] Update README
- [x] Check ESlint vs JShint
- [x] Update dependencies

### CODE REVIEW

**Comment by rogierslag** [comment]
@inventid/tech This might have gone slightly out of hand. Sorry for that!

Anyhow: I really wanted to ES6 this since its the only location in stack where we were using Javascript like it 1995 and that really bothered me. Apart from that this allows us to do later refactorings better.

While I was add it, I improved Docker, ditched Vagrant, removed JSHint, finished eslint and updated all dependencies

Ready for review!

File: index.js [hunk 1]
```diff
@@ -204,15 +223,15 @@ const image = {
 
           im(temp_path)
             .autoOrient()
-            .write(destination_path, function (err) {
+            .write(destination_path, (err) => {
               if (err) {
                 return helpers.send500(res, err);
```
Comment (rogierslag) [inline index.js]: @TiddoLangerak @joostverdoorn is there anything logically to do here that im missing?

File: index.js [hunk 2]
```diff
@@ -188,6 +206,7 @@ const image = {
 
     // We support the file type
     token.consume(sentToken, matches[1], function (err) {
+        // OK #noKidding: Node-sqlite3 bind the `this`, so ES6 syntax is not possible here
         if (err || this.changes !== 1) {
```
Comment (rogierslag) [inline index.js]: @TiddoLangerak @joostverdoorn is there anything logically to do here that im missing?

Comment (joostverdoorn) [inline index.js]: `consume` is your own method. You could make it such that it accepts a callback that accepts the result as its second argument. And yes, node-sqlite3 handles this weirdly as is described in their help.

```javascript
consume(token, id, callback) {
  consumeToken.run([token, id], function(err) {
    callback(err, this);
  });
}
```

Now you can use lamdas in any function calling consume. Not sure if it's the way to go, but it moves this nonsense somewhat higher up the chain.

File: index.js [hunk 3]
```diff
@@ -236,41 +254,43 @@ const image = {
...
+      };
+      res.writeHead(200, headers);
+      fs.readFile(file, (err, data) => {
+        res.end(data);
+      });
+    });
   }
 };
 
 function correctlyResize(file, params, callback) {
```
Comment (joostverdoorn) [inline index.js]: So yea.. Why is this called `correctlyResize` again? Would I want to incorrectly resize?

Comment (rogierslag) [inline index.js]: It used to be done at several places but as you can see in the logic it really isnt that straight forward anymore

File: index.js (hunk 2 above)
Comment (TiddoLangerak) [inline index.js]: Keep in mind though that with the suggestion of @joostverdoorn `this` will no longer be available for non-lambda callbacks as well. You could support this by using `callback.call` or `callback.apply` if you like to.

Comment (rogierslag) [inline index.js]: @TiddoLangerak I'm fine with that, we use it at only one place and I do not expect we will need that somewhere in the future :tm:

I love JS but this magic makes it sometimes less fun than drowning :|

**Comment by rogierslag** [comment]
Ok some more final changes (removed _all_ vagrant references and fixed the `this` thingy. For me this is ready for merge and release :)

**Comment by rogierslag** [comment]
@inventid/tech please....

puppy eyes

File: Token.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      } else {
+        return res.writeHead(403, 'Forbidden').json({error: 'The requested image_id is already requested'}).end();
+      }
+    });
+  },
+  shouldRunCleanup() {
+    return Math.floor(Math.random() * 10) === 0;
+  },
+  cleanup() {
+    log.log('info', 'Doing a token cleanup');
```
Comment (alexnederlof) [inline Token.js]: stringyfied log levels? hahahaha. Node makes me laugh.

File: UrlParsing.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      fileType: matches[2].toLowerCase()
+    };
+  }
+
+  return res;
+}
+};
+
+function splitResizableUrl(url) {
+  return url.match(/^\/(.*)_(\d+)_(\d+)(_(\d+)x)?\.([^?]*)(\?fit=(.*))?$/);
```
Comment (alexnederlof) [inline UrlParsing.js]: if you want other humans to work with your code, break this up into variables next time with logical names. That makes the regex easy to understand.

**Comment by alexnederlof** [comment]
Looks good but I'm no node expert

File: Token.js (hunk 4 above)
Comment (rogierslag) [inline Token.js]: Nah this was just me. Comparable to `System.out.println` and being too lazy to load a logging framework :wink:

File: Dockerfile [hunk 6]
```diff
@@ -6,7 +6,7 @@ RUN apt-get update && \
     apt-get autoremove -y && \
     apt-get clean
 
-RUN npm install -g pm2
+RUN npm install -g pm2 babel-cli babel-preset-es2015
```
Comment (TiddoLangerak) [inline Dockerfile]: Minor thing: but in general it's better to keep the presets in the devDependencies of your package, such that they will just be installed with `npm install` for anyone who wants to contribute to the project.

File: Token.js [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      callback(err, this);
+    });
+  },
+  create(req, res) {
+    // Here we create a token which is valid for one single upload
+    // This way we can directly send the file here and just a small json payload to the app
+    const newToken = uuid.v4();
+    if (!req.body.id) {
+      res.writeHead(400, 'Bad request');
+      console.log(req.body);
```
Comment (TiddoLangerak) [inline Token.js]: This looks like left over debugging

File: UrlParsing.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    matches = splitOriginalUrl(req.url);
+    res = {
+      fileName: matches[1],
+      fileType: matches[2].toLowerCase()
+    };
+  }
+
+  return res;
+}
+};
```
Comment (TiddoLangerak) [inline UrlParsing.js]: The indentation of the object is a bit odd here.

File: index.js [hunk 9]
```diff
@@ -236,41 +253,43 @@ const image = {
...
-        fs.readFile(file, function (err, data) {
-          res.end(data);
-        });
-      });
+    log.log('info', `Requested original image ${matches.fileName} in format ${matches.fileType}`);
+    if (!parsing.supportedFileType(matches.fileType)) {
+      return helpers.set404(res, req.url);
     }
+    const file = `${config.get('originals_dir')}/${matches.fileName}`;
+    fs.exists(file, (exists) => {
```
Comment (TiddoLangerak) [inline index.js]: FYI: `fs.exists` is deprecated and replaced with [`fs.access`](https://nodejs.org/api/fs.html#fs_fs_access_path_mode_callback) (It works a bit different from exists though).

**Comment by TiddoLangerak** [comment]
:+1:

File: README.md [hunk 10]
```diff
@@ -34,10 +34,11 @@ Logging takes place in a JSON Logstash enabled format, so it's easy to get into
 
 ### Requesting an image
 
-You can simply request an image using a plain GET request to `http://localhost:2337/example_100_100.png`.
+You can simply request an image using a plain GET request to `http://localhost:1337/example_100_100.png`.
```
Comment (wouter-willems) [inline README.md]: lol, nice change

File: Token.js [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+      return res.end();
+    }
+    // Ensure the id wasnt requested or used previously
+    insertToken.run([newToken, req.body.id], (err) => {
```
Comment (wouter-willems) [inline Token.js]: FYI the `()` are not needed here

File: UrlParsing.js [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+      return 'image/png';
+    default:
+      return null;
+  }
+},
+  isValidRequest(url) {
+  return splitResizableUrl(url) !== null || splitOriginalUrl(url) !== null;
+},
+  getImageParams(req) {
+  let matches = splitResizableUrl(req.url);
```
Comment (wouter-willems) [inline UrlParsing.js]: this can be const

File: UrlParsing.js [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+  if (matches !== null) {
+    res = {
+      fileName: matches[1],
+      resolutionX: parseInt(matches[2], 10),
+      resolutionY: parseInt(matches[3], 10),
+      fileType: matches[6].toLowerCase(),
+      fit: 'clip'
+    };
+
+    if ( matches[8] && ["clip","crop"].indexOf(matches[8].toLowerCase()) > -1) {
```
Comment (wouter-willems) [inline UrlParsing.js]: matches[8]? kind of confusing what that contains

File: UrlParsing.js [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+      res.fit = matches[8].toLowerCase();
+    }
+
+    if (matches[5] !== undefined) {
```
Comment (wouter-willems) [inline UrlParsing.js]: some here, what is matches[5]? Use a simple comment

File: UrlParsing.js (hunk 5 above)
Comment (wouter-willems) [inline UrlParsing.js]: I agree on this. I prefer some example what this would match

File: index.js [hunk 15]
```diff
@@ -236,41 +253,43 @@ const image = {
...
+      if (!exists) {
+        log.log('warn', 'Image ${matches.fileName} is not available locally');
+        return helpers.send404(res, file);
+      }
+      let headers = {
```
Comment (wouter-willems) [inline index.js]: const

File: index.js [hunk 16]
```diff
@@ -236,41 +253,43 @@ const image = {
...
-    var resizeFactor;
-    var cropX = 0;
-    var cropY = 0;
-    var cropWidth = size.width;
-    var cropHeight = size.height;
+    let resizeFactor;
+    let cropX = 0;
+    let cropY = 0;
+    let cropWidth = size.width;
+    let cropHeight = size.height;
```
Comment (wouter-willems) [inline index.js]: i think these last 2 can be const

**Comment by wouter-willems** [comment]
:+1:

File: index.js (hunk 16 above)
Comment (rogierslag) [inline index.js]: nope, they can be overwritten below

File: README.md (hunk 10 above)
Comment (rogierslag) [inline README.md]: I ditched the Vagrant machine, so now all ports refer to the actual port ;)

**Pull request merged**